core/src/main/scala/kafka/log/Log.scala

```diff
@@ -892,64 +914,70 @@ class Log(@volatile var dir: File,
+      while (segmentEntry != null) {
+        val segment = segmentEntry.getValue
+
+        // If the fetch occurs on the active segment, there might be a race condition where two fetch requests occur after
+        // the message is appended but before the nextOffsetMetadata is updated. In that case the second fetch may
+        // cause OffsetOutOfRangeException. To solve that, we cap the reading up to exposed position instead of the log
+        // end of the active segment.
+        val maxPosition = {
+          if (segmentEntry == segments.lastEntry) {
+            val exposedPos = nextOffsetMetadata.relativePositionInSegment.toLong
+            // Check the segment again in case a new segment has just rolled out.
+            if (segmentEntry != segments.lastEntry)
             // New log segment has rolled out, we can read up to the file end.
+              segment.size
+            else
+              exposedPos
+          } else {
             segment.size
-          else
-            exposedPos
-        } else {
-          segment.size
+          }
         }
-      }
-      val fetchInfo = segment.read(startOffset, maxOffset, maxLength, maxPosition, minOneMessage)
-      if (fetchInfo == null) {
-        segmentEntry = segments.higherEntry(segmentEntry.getKey)
-      } else {
-        return isolationLevel match {
-          case IsolationLevel.READ_UNCOMMITTED => fetchInfo
-          case IsolationLevel.READ_COMMITTED => addAbortedTransactions(startOffset, segmentEntry, fetchInfo)
+        val fetchInfo = segment.read(startOffset, maxOffset, maxLength, maxPosition, minOneMessage)
+        if (fetchInfo == null) {
+          segmentEntry = segments.higherEntry(segmentEntry.getKey)
+        } else {
+          return isolationLevel match {
+            case IsolationLevel.READ_UNCOMMITTED => fetchInfo
+            case IsolationLevel.READ_COMMITTED => addAbortedTransactions(startOffset, segmentEntry, fetchInfo)
+          }
         }
       }
-    }
 
-    // okay we are beyond the end of the last segment with no data fetched although the start offset is in range,
-    // this can happen when all messages with offset larger than start offsets have been deleted.
-    // In this case, we will return the empty set with log end offset metadata
-    FetchDataInfo(nextOffsetMetadata, MemoryRecords.EMPTY)
+      // okay we are beyond the end of the last segment with no data fetched although the start offset is in range,
+      // this can happen when all messages with offset larger than start offsets have been deleted.
+      // In this case, we will return the empty set with log end offset metadata
+      FetchDataInfo(nextOffsetMetadata, MemoryRecords.EMPTY)
+    } catch {
+      case e: IOException =>
+        logDirFailureChannel.maybeAddLogFailureEvent(dir.getParent)
+        throw new KafkaStorageException(s"Exception while reading from $topicPartition in dir ${dir.getParent}", e)
```

Comment:
You are right. I missed this.. I just have gone over all methods in Log.scala.

I just went over all the methods in Log.java and made sure the following rule is satisfied for any method in this class:

1) If the method is used during log load phase, which should only happen before the broker registers itself in the zookeeper, this method doesn't convert IOException to KafkaStorageException.

2) Otherwise, if the method in incurs I/O operation and may be called by other classes after broker has registered itself (excluding shutdown()), this method should catch IOException, add it to log dir failure channel, and re-throw KafkaStorageException

3) Otherwise, the method either doesn't throw IOException, or it is only called by internal methods in Log that belong to one of the two above categories, we don't need to catch IOException in this method.

I still keep an unnecessary conversion from IOException to KafkaStorageException in `Log.roll()`. It is unnecessary because it is currently only called by methods in Log.scala that will catch IOException. I still keep it there since it is a public method and may be used in the future. It is a bit tedious to go over all method and make sure IOException is caught `iff` it is needed. So I choose to lean towards over-catch it in Log.scala. I can remove it if needed.

I also added the comment `We don't convert IOException to KafkaStorageException in this method because this method may be called before all logs are loaded` to some methods in Log.scala so that future develop can be aware of this.